### PR TITLE
Upgrade bitcoin dependency

### PR DESCRIPTION
**Pull request by jrawsthorne**
This is a breaking change

### CODE REVIEW

File: src/bitcoin_adaptor.rs [hunk 1]
```diff
@@ -45,14 +42,14 @@ impl BitcoinAdaptor {
...
-        Ok(self.hammersbald.put_keyed(&encodable.bitcoin_hash()[..], serde_cbor::to_vec(encodable)?.as_slice())?)
+    pub fn put_hash_keyed<T>(&mut self, id: &sha256d::Hash, encodable: &T) -> Result<PRef, Box<dyn Error>>
+        where T: Serialize {
+        Ok(self.hammersbald.put_keyed(&id[..], serde_cbor::to_vec(encodable)?.as_slice())?)
     }
 
     /// Retrieve a bitcoin_object with its hash
     pub fn get_hash_keyed<T>(&self, id: &sha256d::Hash) -> Result<Option<(PRef, T)>, Box<dyn Error>>
-        where T: DeserializeOwned + BitcoinHash{
+        where T: DeserializeOwned {
```
Comment (erasmospunk) [inline src/bitcoin_adaptor.rs:52]: Why do we need the `BitcoinHash` removed here?

File: src/bitcoin_adaptor.rs [hunk 2]
```diff
@@ -201,9 +198,9 @@ mod test {
 
         let genesis = genesis_block(Network::Bitcoin);
         // store the genesist block
-        bdb.put_hash_keyed(&genesis).unwrap();
+        bdb.put_hash_keyed(&genesis.bitcoin_hash().as_hash(), &genesis).unwrap();
         // find it
-        if let Some((_, block)) = bdb.get_hash_keyed::<Block>(&genesis.bitcoin_hash()).unwrap() {
+        if let Some((_, block)) = bdb.get_hash_keyed::<Block>(&genesis.bitcoin_hash().as_hash()).unwrap() {
```
Comment (erasmospunk) [inline src/bitcoin_adaptor.rs:203]: Maybe use `block_hash` instead?

File: src/bitcoin_adaptor.rs [hunk 3]
```diff
@@ -45,14 +42,14 @@ impl BitcoinAdaptor {
     }
 
     /// Store some bitcoin object that has a bitcoin hash
-    pub fn put_hash_keyed<T>(&mut self, encodable: &T) -> Result<PRef, Box<dyn Error>>
-        where T: Serialize + BitcoinHash {
-        Ok(self.hammersbald.put_keyed(&encodable.bitcoin_hash()[..], serde_cbor::to_vec(encodable)?.as_slice())?)
+    pub fn put_hash_keyed<T>(&mut self, id: &sha256d::Hash, encodable: &T) -> Result<PRef, Box<dyn Error>>
```
[the remark this answers is not in the document]
Comment (erasmospunk) [reply src/bitcoin_adaptor.rs:45]: Maybe T could implement an `Into<Hash>` here?

**Review by erasmospunk — CHANGES_REQUESTED** [review]

File: src/bitcoin_adaptor.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (jrawsthorne) [reply src/bitcoin_adaptor.rs:45]: My reasoning was that the BitcoinHash trait is being removed and sha256d::Hash is hard coded for get so it makes sense that you can only insert with sha256d::Hash

File: src/bitcoin_adaptor.rs (hunk 1 above)
Comment (jrawsthorne) [reply src/bitcoin_adaptor.rs:52]: See above, this will be removed in the next version

File: src/bitcoin_adaptor.rs (hunk 2 above)
Comment (jrawsthorne) [reply src/bitcoin_adaptor.rs:203]: block_hash is only available on master at the moment, got to wait for the next rust-bitcoin release

**Comment by jrawsthorne** [comment]
It's still a bit awkward with get_hash_keyed because of the extra generic. The other way is to take a sha256d::Hash but then the caller has to call as_hash(). Probably not worth it to discuss too much as this is only really to get murmel to compile and hammersbald probably won't be used in the future

File: src/bitcoin_adaptor.rs (hunk 1 above)
Comment (notmandatory) [reply src/bitcoin_adaptor.rs:52]: @jrawsthorne I agree this makes the api easier to use. But I'm not as familiar with how the generic Into<> works. So I manually ran the tests (with `cargo test --features="bitcoin_support"`). The existing tests confirms the values passed in for the `id` argument is borrowed and can still be accessed after being passed into the get_hash_keyed function. So looks good to me.

File: src/bitcoin_adaptor.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (notmandatory) [reply src/bitcoin_adaptor.rs]: One way to keep the same API would be to add the `BitcoinHash` trait to the chaindb module and implement it on `StoredHeader`, which is the only type passed to push_hash_keyed. It's hard to know if any other projects are using this lib so I'm sympathetic to not changing the API too much if we don't have to.

**Comment by notmandatory** [comment]
Changes look good to me. I also confirmed all tests pass: 
```
cargo test --features="bitcoin_support"
```

**Review by dr-orlovsky — CHANGES_REQUESTED** [review]
NACK: BitcoinHash trait was deliberately removed from rust-bitcoin and replaced with normal txid/wtxid() and block_hash functions. I think it will be better to adopt this new API and not just to introduce deprecated trait

**Comment by notmandatory** [comment]
@jrawsthorne it sounds like your original approach at commit on the 25th with the API change and no BitcoinHash trait is the right way to go. I'll withdraw my suggestion. :-)

File: Cargo.toml [hunk 4]
```diff
@@ -22,7 +22,7 @@ rand="0.7"
 byteorder="1.2"
 lru-cache = "0.1.1"
 bitcoin_hashes = "0.7"
-bitcoin = { version= "0.21", features=["serde"], optional = true }
+bitcoin = { git = "https://github.com/rust-bitcoin/rust-bitcoin", rev="e8bcde4", features=["use-serde"], optional = true }
```
Comment (stevenroose) [inline Cargo.toml:25]: Why is this using a git branch instead of a release?

**Review by stevenroose — CHANGES_REQUESTED** [review]

File: Cargo.toml (hunk 4 above)
Comment (notmandatory) [reply Cargo.toml:25]: I think the reason is that at the time this PR was create there wasn't a recent tag available, but I see `0.24.0` was just tagged a couple weeks ago, so probably should be using that one now.

**Comment by jrawsthorne** [comment]
Reverted to the previous state and updated the bitcoin dependency to 0.25 rather than git

File: Cargo.toml [hunk 5]
```diff
@@ -19,12 +19,12 @@ bitcoin_support=["bitcoin","serde","serde_cbor"]
...
+lru-cache = "0.1.2"
+bitcoin_hashes = "0.9"
+bitcoin = { version = "0.25", features=["use-serde"], optional = true }
 serde = { version ="1", optional = true }
-serde_cbor = { version="0.10", optional = true }
+serde_cbor = { version="0.11", optional = true }
 
 [dev-dependencies]
-hex = "0.3"
+hex = "0.4"
```
Comment (dr-orlovsky) [inline Cargo.toml:30]: You can get rid of this dependency like it was done for rust-bitcoin (https://github.com/rust-bitcoin/rust-bitcoin/pull/482/commits/07fb14b8872507c40e483c5524e72067c372f2db) and use `Vec::from_hex` instead

**Review by dr-orlovsky — APPROVED** [review]
utACK https://github.com/rust-bitcoin/hammersbald/pull/13/commits/b753e9dcb725a7ed0a1863076b386f09d20c46ad

My comment on removing hex dev dependency can go into a separate PR

**Comment by stevenroose** [comment]
Closed in favor of https://github.com/rust-bitcoin/hammersbald/pull/14.

**Pull request closed without merging**